src/transformers/generation/utils.py

```diff
@@ -2001,11 +2001,26 @@ def _prepare_cache_for_generation(
         # Use DynamicCache() instance by default. This will avoid back and forth from legacy format that
         # keeps copying the cache thus using much more memory
         else:
-            model_kwargs[cache_name] = (
-                DynamicCache(**dynamic_cache_kwargs)
-                if not requires_cross_attention_cache
-                else EncoderDecoderCache(DynamicCache(**dynamic_cache_kwargs), DynamicCache(**dynamic_cache_kwargs))
-            )
+            if not requires_cross_attention_cache:
+                model_kwargs[cache_name] = DynamicCache(**dynamic_cache_kwargs)
+            else:
+                # For encoder-decoder models, we need to use separate configs for encoder and decoder
+                decoder_cache_kwargs = dynamic_cache_kwargs.copy()
+                encoder_cache_kwargs = dynamic_cache_kwargs.copy()
```

Comment:
`dynamic_cache_kwargs` only contains a `config` which we know for sure we won't use in this branch -- we don't need to copy the object, we can create a new dictionary